docs(event_handler): improve routing rules syntax documentation

dap0am opened the pull request:
**Issue number:** #2962 

Fixes #2962

## Summary

This PR addresses issue #2962 by adding comprehensive documentation for routing rules syntax in the Event Handler.

### Changes

- Added detailed "Routing Rules Syntax" section explaining:
  - Dynamic path parameter syntax (`<parameter_name>`)
  - Supported characters for path parameters (including special chars like `-._~()'\\!*:@,;=+&$%<> \[]{}|^`)
  - How routes are converted to regex patterns behind the scenes
  - Advanced examples with multiple parameters
  - Complex routing scenarios

- Enhanced "Catch-all routes" section with:
  - Detailed regex pattern examples
  - Common use cases for proxy routes and file paths
  - Route matching priority explanation
  - Mixed dynamic parameter + regex examples

### User experience

**Before:** Developers had no documentation about routing rules syntax, leading to confusion about:
- How the `<parameter>` syntax works
- What regex patterns are supported
- What characters are allowed in path parameters
- How route matching priority works

**After:** Developers now have comprehensive documentation with:
- Clear explanation of dynamic path parameter syntax
- Table showing supported characters in path parameters
- Examples of complex routing patterns
- Regex pattern conversion details
- Route matching behavior and priority rules

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://docs.powertools.aws.dev/lambda/python/#tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/aws-powertools/powertools-lambda-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**: N/A

This is a documentation-only change with no breaking changes.

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

dap0am committed:
docs: add comprehensive routing rules syntax documentation

Addresses issue #2962 by adding detailed documentation for:
- Dynamic path parameter syntax (<parameter_name>)
- Supported characters in path parameters
- Regex pattern conversion process
- Advanced routing examples with complex parameters
- Catch-all routes with regex patterns
- Route matching priority and behavior

The new documentation provides clear explanations of how routing rules work,
the regex syntax supported, and examples of complex routing patterns that
were previously undocumented.

leandrodamascena commented on docs/core/event_handler/api_gateway.md:
```diff
@@ -214,17 +214,145 @@ Each dynamic route you set must be part of your function signature. This allows
 
+#### Routing Rules Syntax
+
+The routing system uses a specific syntax to define dynamic URL patterns. Understanding this syntax is crucial for creating flexible and robust API routes.
```
I think we can remove this entire section. It doesn't add much relevant information.

leandrodamascena commented on docs/core/event_handler/api_gateway.md:
```diff
@@ -214,17 +214,145 @@ Each dynamic route you set must be part of your function signature. This allows
+
+The routing system uses a specific syntax to define dynamic URL patterns. Understanding this syntax is crucial for creating flexible and robust API routes.
+
+##### Dynamic Path Parameters
```
```suggestion
#### Dynamic Path Parameters
```

leandrodamascena commented on docs/core/event_handler/api_gateway.md:
```diff
@@ -214,17 +214,145 @@ Each dynamic route you set must be part of your function signature. This allows
+
+##### Dynamic Path Parameters
+
+Dynamic path parameters are defined using angle brackets `<parameter_name>` syntax. These parameters are automatically converted to regex patterns for efficient route matching.
```
```suggestion
Dynamic path parameters are defined using angle brackets `<parameter_name>` syntax. These parameters are automatically converted to regex patterns for efficient route matching and performance gains.
```

leandrodamascena commented on docs/core/event_handler/api_gateway.md:
```diff
@@ -214,17 +214,145 @@ Each dynamic route you set must be part of your function signature. This allows
+**Syntax**: `/path/<parameter_name>`
+
+* **Parameter names** must contain only word characters (letters, numbers, underscore)
+* **Captured values** can contain letters, numbers, underscores, and these special characters: `-._~()'!*:@,;=+&$%<> \[]{}|^`
```
I really like this! 👍

leandrodamascena commented on docs/core/event_handler/api_gateway.md:
````diff
@@ -214,17 +214,145 @@ Each dynamic route you set must be part of your function signature. This allows
+* **Parameter names** must contain only word characters (letters, numbers, underscore)
+* **Captured values** can contain letters, numbers, underscores, and these special characters: `-._~()'!*:@,;=+&$%<> \[]{}|^`
+
+=== "routing_syntax_basic.py"
+
+    ```python
+    from aws_lambda_powertools.event_handler import APIGatewayRestResolver
+
+    app = APIGatewayRestResolver()
+
+    @app.get("/users/<user_id>")
+    def get_user(user_id: str):
+        # user_id can be: "123", "user-456", "john.doe", "user_with_underscores"
+        return {"user_id": user_id}
+
+    @app.get("/orders/<order_id>/items/<item_id>")
+    def get_order_item(order_id: str, item_id: str):
+        # Multiple parameters: /orders/ORD-123/items/ITEM_456
+        return {"order_id": order_id, "item_id": item_id}
+    ```
````
Could you create a file in the `examples/event_handler_rest/src` folder, move this example there, and refer to it here? We don't add inline examples because we want to ensure our examples are accurate and we can pass linter and formats in those files.

Also, this example is missing the Lambda entrypoint, aka `lambda handler`. please check other examples and use `LambdaContext` to annotate the context parameter.

leandrodamascena commented on docs/core/event_handler/api_gateway.md:
```diff
@@ -214,17 +214,145 @@ Each dynamic route you set must be part of your function signature. This allows
+|---------------|-----------------|---------|---------------|
+| `/users/<user_id>` | `^/users/(?P<user_id>[-._~()'!*:@,;=+&$%<> \[\]{}|^\w]+)$` | `/users/123`, `/users/user-456` | `/users/123/profile` |
+| `/api/<version>/users` | `^/api/(?P<version>[-._~()'!*:@,;=+&$%<> \[\]{}|^\w]+)/users$` | `/api/v1/users`, `/api/2.0/users` | `/api/users` |
+| `/files/<path>` | `^/files/(?P<path>[-._~()'!*:@,;=+&$%<> \[\]{}|^\w]+)$` | `/files/document.pdf`, `/files/folder%20name` | `/files/sub/folder/file.txt` |
```
Please move this table to Dynamic Path Parameters section.. this make more sense there.

leandrodamascena commented on docs/core/event_handler/api_gateway.md:
```diff
@@ -214,17 +214,145 @@ Each dynamic route you set must be part of your function signature. This allows
+
+##### Regex Pattern Conversion
+
+Behind the scenes, dynamic routes are converted to regex patterns for efficient matching:
```
After moving the table you can remove this section.

leandrodamascena commented on docs/core/event_handler/api_gateway.md:
```diff
@@ -214,17 +214,145 @@ Each dynamic route you set must be part of your function signature. This allows
+???+ warning "Route Matching Behavior"
+    * Routes are matched **exactly** - no partial matches
+    * Dynamic parameters match **non-slash characters only** by default
+    * For paths with slashes, use [catch-all routes](#catch-all-routes) instead
```
Please remove this.

leandrodamascena commented on docs/core/event_handler/api_gateway.md:
```diff
@@ -214,17 +214,145 @@ Each dynamic route you set must be part of your function signature. This allows
+    * Dynamic parameters match **non-slash characters only** by default
+    * For paths with slashes, use [catch-all routes](#catch-all-routes) instead
+
+##### Advanced Examples
```
```suggestion
#### Advanced Examples
```

leandrodamascena commented on docs/core/event_handler/api_gateway.md:
````diff
@@ -214,17 +214,145 @@ Each dynamic route you set must be part of your function signature. This allows
+    # /files/my-file_v2.txt → filename = "my-file_v2.txt"
+    # /files/file%20with%20spaces → filename = "file%20with%20spaces"
+    return {"filename": filename}
+```
````
Could you create files in the examples/event_handler_rest/src folder, move these example there, and refer to it here? We don't add inline examples because we want to ensure our examples are accurate and we can pass linter and formats in those files.

Also, these examples are missing the Lambda entrypoint, aka lambda handler. please check other examples and use LambdaContext to annotate the context parameter.

leandrodamascena commented on docs/core/event_handler/api_gateway.md:
````diff
@@ -214,17 +214,145 @@ Each dynamic route you set must be part of your function signature. This allows
+```
+
+???+ tip "Function Parameter Names Must Match"
+    The parameter names in your route (`<user_id>`) must exactly match the parameter names in your function signature (`user_id: str`). This is how the framework knows which captured values to pass to which parameters.
````
```suggestion
???+ tip "Function parameter names must match"
    The parameter names in your route (`<user_id>`) must exactly match the parameter names in your function signature (`user_id: str`). This is how the framework knows which captured values to pass to which parameters.
```

leandrodamascena reviewed (changes requested):
Hey @dap0am thanks a lot for working in this PR! I left some comments before we move to the next round of review.

Please let me know if you have any question.

dap0am commented:
> Hey @dap0am thanks a lot for working in this PR! I left some comments before we move to the next round of review.
> 
> Please let me know if you have any question.

Thanks for the comment @leandrodamascena , will work on them as soon as possible.

dap0am committed:
docs: refactor routing rules syntax documentation per review feedback

- Remove redundant routing rules syntax introduction section
- Update header levels from h5 to h4 for better hierarchy
- Move inline code examples to separate files in examples/event_handler_rest/src
- Add proper Lambda handlers with Logger and Tracer utilities
- Reorganize content by moving regex pattern table to Dynamic Path Parameters section
- Update tip formatting to match project style
- Simplify inline comments to follow project conventions

dap0am commented:
Hi @leandrodamascena all comments has been actioned, please review and let me know if there is anything that I missed. Thank you.

leandrodamascena committed:
Small changes

leandrodamascena committed:
Merge branch 'develop' into docs/routing-rules-documentation

leandrodamascena reviewed (approved):
Hi @dap0am, thank you so much for working on this PR! I just made a few minor final changes to follow some documentation guidelines we have, but nothing you missed!

It was great working with you on this PR! I'm opening more issues that might be interesting to work on. I hope to see you again here.

APPROVED!

dap0am commented:
> Hi @dap0am, thank you so much for working on this PR! I just made a few minor final changes to follow some documentation guidelines we have, but nothing you missed!
> 
> It was great working with you on this PR! I'm opening more issues that might be interesting to work on. I hope to see you again here.
> 
> APPROVED!

No worries @leandrodamascena , will keep an eye on the issues to see if I fancy anyone.
